controllers/tykoasapidefinition_controller.go

```diff
@@ -188,6 +224,192 @@ func (r *TykOasApiDefinitionReconciler) Reconcile(ctx context.Context, req ctrl.
 	return ctrl.Result{RequeueAfter: reqA}, err
 }
 
+// reconcileVersionedOasApi checks if a versioned API can be deleted and isn't a dangling Versioned API
+// i.e an API that was previously part of the version list of a Base API.
+func (r *TykOasApiDefinitionReconciler) reconcileVersionedOasApi(
+	ctx context.Context,
+	log logr.Logger,
+	versionedTykOAS *v1alpha1.TykOasApiDefinition,
+) error {
+	if err := r.cleanUpDanglingVersionedOasAPIs(ctx, versionedTykOAS, log); err != nil {
+		return err
+	}
+
+	if !versionedTykOAS.ObjectMeta.DeletionTimestamp.IsZero() {
+		if versionedTykOAS.Status.GetIsDefaultVersion() {
+			errMsg := "Cannot delete default API version. Remove it from base API's default first"
+
+			versionedTykOAS.Status.LatestTransaction = v1alpha1.TransactionInfo{
+				Time:   metav1.Now(),
+				Status: v1alpha1.TransactionStatus(errMsg),
+			}
+
+			log.Info("error deleting versioned TykOasAPI", errMsg)
+
+			if err := r.Status().Update(ctx, versionedTykOAS); err != nil {
+				return err
+			}
+
+			return nil
+		}
+
+		var baseOasAPI v1alpha1.TykOasApiDefinition
+
+		if err := r.Get(ctx, types.NamespacedName{
+			Name:      versionedTykOAS.Status.GetBaseVersionName(),
+			Namespace: versionedTykOAS.Status.GetBaseVersionNamespace(),
+		}, &baseOasAPI); err != nil {
+			if k8serrors.IsNotFound(err) {
+				log.Info("failed to get TykOasApiDefinition CR", "err", err.Error())
+			}
+
+			return err
+		}
+
+		errMsg := "Cannot delete API. It is referenced as version in another Tyk OAS API. Please remove it before deleting."
+
+		versionedTykOAS.Status.LatestTransaction = v1alpha1.TransactionInfo{
+			Time:   metav1.Now(),
+			Status: v1alpha1.TransactionStatus(errMsg),
+		}
+
+		log.Info("error deleting versioned TykOasAPI", errMsg)
+
+		if err := r.Status().Update(ctx, versionedTykOAS); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *TykOasApiDefinitionReconciler) cleanUpDanglingVersionedOasAPIs(
+	ctx context.Context,
+	versionedTykOAS *v1alpha1.TykOasApiDefinition,
+	log logr.Logger,
+) error {
+	if versionedTykOAS.Status.GetBaseVersionName() == "" {
+		return nil
+	}
+
+	var baseOasAPI v1alpha1.TykOasApiDefinition
+
+	ns := versionedTykOAS.Status.GetBaseVersionNamespace()
+	if ns == "" {
+		ns = "default"
```

Comment:
any specific reason to use `"default"` namespace as a default value? should we use the namespace of `versionedTykOAS`? What do you think?